### PR TITLE
Controller sse firehose

### PR DESCRIPTION
**Pull request by jvatic**
#737

### CODE REVIEW

File: controller/app.go [hunk 1]
```diff
@@ -191,69 +209,39 @@ func (r *AppRepo) List() (interface{}, error) {
...
-	row := r.db.QueryRow("SELECT r.release_id, r.artifact_id, r.data, r.created_at FROM apps a JOIN releases r USING (release_id) WHERE a.app_id = $1", id)
-	return scanRelease(row)
+	e.App = app
+	data, err := json.Marshal(e)
+	if err != nil {
+		fmt.Printf("AppRepo.createEvent error (json.Marshal): %#v\n", err)
+		return err
+	}
+	query := "INSERT INTO events (app_id, object_id, object_type, data) VALUES ($1, $2, $3, $4)"
+	if err := r.db.Exec(query, appID, appID, string(ct.EventTypeApp), data); err != nil {
```
Comment (lmars) [inline controller/app.go]: Given these events all follow DB actions, I wonder whether we should be creating the events using DB triggers (e.g. [like this](https://github.com/flynn/flynn/blob/master/controller/schema.go#L80))?

**Comment by jvatic** [comment]
The postgres `to_json` function doesn't format time fields correctly, going back to creating events from within Go

\\cc @lmars

**Comment by jvatic** [comment]
Using postgres 9.4 for unit tests fixes the formatting of time fields, however there are a few other factors which make creating events in the Go code preferable over db triggers when JSON encoding in involved:

- Some fields have different names in the db vs JSON
- Some fields should be omitted when null

**Comment by jvatic** [comment]
ping @lmars, @titanous this is ready for a review pass

File: controller/app.go [hunk 2]
```diff
@@ -171,11 +176,17 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
 		}
 	}
 
-	return app, tx.Commit()
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+
+	r.createAppEvent(app.ID)
```
Comment (lmars) [inline controller/app.go]: Should we not be creating the app events inside the transaction, and throwing an error if it fails?

File: controller/client/client.go [hunk 3]
```diff
@@ -486,9 +486,9 @@ outer:
 
 // StreamJobEvents streams job events to the output channel.
 func (c *Client) StreamJobEvents(appID string, output chan *ct.JobEvent) (stream.Stream, error) {
-	appEvents := make(chan *ct.AppEvent)
-	go convertAppEvents(appEvents, output)
-	return c.ResumingStream("GET", fmt.Sprintf("/apps/%s/events?object_type=%s", appID, ct.EventTypeJob), appEvents)
+	appEvents := make(chan *ct.Event)
+	go convertEvents(appEvents, output)
+	return c.ResumingStream("GET", fmt.Sprintf("/apps/%s/events?object_types=%s", appID, ct.EventTypeJob), appEvents)
```
Comment (lmars) [inline controller/client/client.go]: We should probably just move to a single path for events and just return `StreamEvents({AppID: appID, ObjectTypes: [ct.EventTypeJob]}` here

File: controller/events.go [hunk 4]
```diff
@@ -1,16 +1,199 @@
 package main
 
 import (
+	"database/sql"
```
Comment (lmars) [inline controller/events.go]: This import should be `github.com/flynn/flynn/Godeps/_workspace/src/github.com/flynn/go-sql`

File: controller/events.go [hunk 5]
```diff
@@ -177,6 +370,13 @@ func (e *EventListener) Notify(event *ct.AppEvent) {
 			sub.Notify(event)
 		}
 	}
+	if event.AppID != "" {
```
Comment (lmars) [inline controller/events.go]: Should this be inverted?

File: controller/events_test.go [hunk 6]
```diff
@@ -78,3 +81,273 @@ func (s *S) TestAppEvents(c *C) {
...
+			var eventApp *ct.App
+			c.Assert(json.Unmarshal(e.Data, &eventApp), IsNil)
+			c.Assert(e.AppID, Equals, app.ID)
+			c.Assert(e.ObjectType, Equals, ct.EventTypeApp)
+			c.Assert(e.ObjectID, Equals, app.ID)
+			c.Assert(eventApp, NotNil)
+			c.Assert(eventApp.ID, Equals, app.ID)
+			fn(eventApp)
+		case <-time.After(time.Second):
+			c.Fatal(fmt.Sprintf("Timed out waiting for event %d", i))
```
Comment (lmars) [inline controller/events_test.go]: Use `c.Fatalf`

File: controller/events_test.go [hunk 7]
```diff
@@ -78,3 +81,273 @@ func (s *S) TestAppEvents(c *C) {
...
+			}
+			var eventApp *ct.App
+			c.Assert(json.Unmarshal(e.Data, &eventApp), IsNil)
+			c.Assert(e.AppID, Equals, app.ID)
+			c.Assert(e.ObjectType, Equals, ct.EventTypeApp)
+			c.Assert(e.ObjectID, Equals, app.ID)
+			c.Assert(eventApp, NotNil)
+			c.Assert(eventApp.ID, Equals, app.ID)
+			fn(eventApp)
+		case <-time.After(time.Second):
```
Comment (lmars) [inline controller/events_test.go]: We should probably up this timeout to 10s, CI can get slow sometimes.

File: controller/formation.go [hunk 8]
```diff
@@ -138,9 +141,29 @@ func (r *FormationRepo) Remove(appID, releaseID string) error {
...
+func (r *FormationRepo) createScaleEvent(formation *ct.Formation) {
+	logErr := func(err error) {
+		log.Printf("Error creating scale event: %s\n", err)
+	}
+	data, err := json.Marshal(formation.Processes)
+	if err != nil {
+		logErr(err)
+		return
+	}
+	objectID := strings.Join([]string{formation.AppID, formation.ReleaseID}, ":")
```
Comment (lmars) [inline controller/formation.go]: I'd probably just do `formation.AppID + ":" + formation.ReleaseID`

File: controller/key.go [hunk 9]
```diff
@@ -82,3 +92,25 @@ func (r *KeyRepo) List() (interface{}, error) {
...
+		return
+	}
+	query := "INSERT INTO events (object_id, object_type, data) VALUES ($1, $2, $3)"
+	if err := r.db.Exec(query, key.ID, string(ct.EventTypeKey), data); err != nil {
+		logErr(err)
+	}
+}
+
+func (r *KeyRepo) createKeyDeletionEvent(keyID string) {
+	query := "INSERT INTO events (object_id, object_type) VALUES ($1, $2)"
```
Comment (lmars) [inline controller/key.go]: Any reason not to include the data of the deleted key?

File: controller/release.go [hunk 10]
```diff
@@ -71,12 +72,33 @@ func (r *ReleaseRepo) Add(data interface{}) error {
...
+	}
+	data, err := json.Marshal(release)
+	if err != nil {
+		logErr(err)
+		return
+	}
+	query := "INSERT INTO events (object_id, object_type, data) VALUES ($1, $2, $3)"
+	if err := r.db.Exec(query, release.ID, string(ct.EventTypeRelease), data); err != nil {
+		logErr(err)
+	}
```
Comment (lmars) [inline controller/release.go]: The event creation methods are all pretty similar, could refactor to a more generic:

```go
func createEvent(objectID string, objectType ct.EventType, v interface{})
```

File: controller/resource.go [hunk 11]
```diff
@@ -173,7 +179,35 @@ func (r *ResourceRepo) Remove(id string) error {
...
+	logErr := func(err error) {
+		log.Printf("Error creating resource event: %s\n", err)
+	}
+	data, err := json.Marshal(resource)
+	if err != nil {
+		logErr(err)
+		return
+	}
+	objectID := strings.Join([]string{resource.ProviderID, resource.ID}, ":")
+	query := "INSERT INTO events (object_id, object_type, data) VALUES ($1, $2, $3)"
```
Comment (lmars) [inline controller/resource.go]: Resources can be associated with many apps, so does it make sense to insert this event for all the associated apps?

File: controller/events.go (hunk 5 above)
Comment (jvatic) [inline controller/events.go]: Nope, this is so subscribers not filtering by app get the event. Added a comment on the next line.

**Comment by jvatic** [comment]
@lmars comments addressed

File: controller/controller.go [hunk 12]
```diff
@@ -305,6 +318,26 @@ func (c *controllerAPI) getRoute(ctx context.Context) (*router.Route, error) {
...
+	if err != nil {
+		return err
+	}
+	if appID == "" {
+		query := "INSERT INTO events (object_id, object_type, data) VALUES ($1, $2, $3)"
+		if _, err := dbExec(query, objectID, string(objectType), encodedData); err != nil {
+			return err
+		}
+
+	} else {
```
Comment (lmars) [inline controller/controller.go]: Return the `err` in the `if` branch and drop the `else`

**Comment by lmars** [comment]
This needs a rebase, ensuring we maintain the fixes from #1667 and #1678.

File: controller/events.go (hunk 5 above)
Comment (lmars) [inline controller/events.go]: Not sure I follow. Let's say I subscribe to artifact changes whose events don't have `AppID` set, my subscriber will be in `subscribers[""]`, and when an artifact event comes along, `event.AppID` will be `""` so I won't get notified due to this condition?

File: controller/app.go [hunk 13]
```diff
@@ -73,21 +79,29 @@ func (r *AppRepo) Add(data interface{}) error {
 			Domain:  fmt.Sprintf("%s.%s", app.Name, r.defaultDomain),
 			Service: app.Name + "-web",
 		}).ToRoute()
-		route.ParentRef = routeParentRef(app.ID)
-		if err := r.router.CreateRoute(route); err != nil {
+		if err := createRoute(r.db, r.router, app.ID, route); err != nil {
```
Comment (lmars) [inline controller/app.go]: We should move this logic to after committing the transaction, so if the transaction fails we don't have a dangling route.

File: controller/routes.go [hunk 14]
```diff
@@ -5,11 +5,32 @@ import (
...
+		return err
+	}
+	if err := rc.CreateRoute(route); err != nil {
+		return err
+	}
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	if err := createEvent(tx.Exec, appID, route.ID, ct.EventTypeRoute, route); err != nil {
```
Comment (lmars) [inline controller/routes.go]: Does this need to be in a transaction?

File: controller/routes.go [hunk 15]
```diff
@@ -65,5 +80,21 @@ func (c *controllerAPI) DeleteRoute(ctx context.Context, w http.ResponseWriter,
 		respondWithError(w, err)
 		return
 	}
+	tx, err := c.appRepo.db.Begin()
+	if err != nil {
+		respondWithError(w, err)
+		return
+	}
+	appID := c.getApp(ctx).ID
+	if err := createEvent(tx.Exec, appID, route.ID, ct.EventTypeRouteDeletion, route); err != nil {
```
Comment (lmars) [inline controller/routes.go]: Does this need to be in a transaction?

File: controller/events.go (hunk 5 above)
Comment (jvatic) [inline controller/events.go]: It's to handle the case where the subscriber is in `subscribers[""]` and `event.AppID` is set, e.g. subscribing to all app events but not specifying an object ID.

File: controller/routes.go (hunk 14 above)
Comment (jvatic) [inline controller/routes.go]: `tx.Exec` returns `(sql.Result, error)` and `db.Exec` just returns `error`, the transaction allows `createEvent` to be simpler. I suppose `db.Exec` could be wrapped in a function matching the `tx.Exec` signature.

File: controller/routes.go (hunk 15 above)
Comment (jvatic) [inline controller/routes.go]: Ditto

File: controller/app.go (hunk 13 above)
Comment (jvatic) [inline controller/app.go]: Done

**Comment by jvatic** [comment]
Route events should be streamed from the router

**Comment by jvatic** [comment]
@lmars comments addressed

I'm refactoring the route events to be streamed from the router.

**Comment by jvatic** [comment]
@lmars ready for another review pass

File: router/api.go [hunk 16]
```diff
@@ -152,3 +155,45 @@ func deleteRoute(params martini.Params, router *Router, r render.Render) {
...
+	tcpEvents := make(chan *router.Event)
+	sseEvents := make(chan *router.StreamEvent)
+	go r.HTTP.Watch(httpEvents)
+	go r.TCP.Watch(tcpEvents)
+	defer r.HTTP.Unwatch(httpEvents)
+	defer r.TCP.Unwatch(tcpEvents)
+	go func() {
+		for {
+			select {
+			case e, ok := <-httpEvents:
```
Comment (jvatic) [inline router/api.go]: Don't need the select

File: router/api.go [hunk 17]
```diff
@@ -152,3 +155,45 @@ func deleteRoute(params martini.Params, router *Router, r render.Render) {
...
+					Route: e.Route,
+					Error: e.Error,
+				}
+			}
+		}
+	}()
+	go func() {
+		for {
+			select {
+			case e, ok := <-tcpEvents:
```
Comment (jvatic) [inline router/api.go]: Ditto

File: router/api.go [hunk 18]
```diff
@@ -152,3 +155,45 @@ func deleteRoute(params martini.Params, router *Router, r render.Render) {
...
+				if !ok {
+					return
+				}
+				sseEvents <- &router.StreamEvent{
+					Event: e.Event,
+					Route: e.Route,
+					Error: e.Error,
+				}
```
Comment (jvatic) [inline router/api.go]: Pull this out into a func as this is identical to above

File: controller/controller.go [hunk 19]
```diff
@@ -105,6 +115,42 @@ func main() {
...
+	if err != nil {
+		return err
+	}
+	go func() {
+		for {
+			e, ok := <-events
+			if !ok {
+				return
+			}
+			route := e.Route
```
Comment (jvatic) [inline controller/controller.go]: `route.ID = postgres.CleanUUID(route.ID)`

File: controller/controller.go [hunk 20]
```diff
@@ -105,6 +115,42 @@ func main() {
...
+			var appID string
+			if strings.HasPrefix(route.ParentRef, routeParentRefPrefix) {
+				appID = strings.TrimPrefix(route.ParentRef, routeParentRefPrefix)
+			}
+			eventType := ct.EventTypeRoute
+			if e.Event == "remove" {
+				eventType = ct.EventTypeRouteDeletion
+			}
+			if err := createEvent(wrapDBExec(db.Exec), appID, route.ID, eventType, route); err != nil {
```
Comment (jvatic) [inline controller/controller.go]: This should probably set `events.unique_id`

File: controller/controller.go [hunk 21]
```diff
@@ -305,6 +364,21 @@ func (c *controllerAPI) getRoute(ctx context.Context) (*router.Route, error) {
 	return route, err
 }
 
+func createEvent(dbExec func(string, ...interface{}) (sql.Result, error), appID, objectID string, objectType ct.EventType, data interface{}) error {
```
Comment (jvatic) [inline controller/controller.go]: Refactor to `func createEvent(dbExec func(string, ...interface{}) (sql.Result, error), e *ct.Event, data interface{}) error`

**Comment by jvatic** [comment]
There should be an integration test for route events

File: controller/controller.go (hunk 19 above)
Comment (jvatic) [inline controller/controller.go]: Done

**Comment by jvatic** [comment]
ping @lmars for review

File: controller/schema.go [hunk 22]
```diff
@@ -42,27 +42,32 @@ func migrateDB(db *sql.DB) error {
...
-		`CREATE FUNCTION notify_app_event() RETURNS TRIGGER AS $$
+
+		`CREATE UNIQUE INDEX ON events (unique_id)`,
+		`CREATE FUNCTION notify_event() RETURNS TRIGGER AS $$
     BEGIN
-	PERFORM pg_notify('app_events', NEW.event_id || ':' || NEW.app_id);
+  IF NEW.app_id IS NOT NULL THEN
+    PERFORM pg_notify('events', NEW.event_id || ':' || NEW.app_id);
+  ELSE
+    PERFORM pg_notify('events', NEW.event_id || '');
```
Comment (titanous) [inline controller/schema.go]: Is the append to an empty string necessary?

File: controller/artifact.go [hunk 23]
```diff
@@ -27,17 +27,38 @@ func (r *ArtifactRepo) Add(data interface{}) error {
...
 	}
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
 	a.ID = postgres.CleanUUID(a.ID)
-	return err
+	if err := createEvent(tx.Exec, &ct.Event{
+		ObjectID:   a.ID,
+		ObjectType: ct.EventTypeArtifact,
```
Comment (titanous) [inline controller/artifact.go]: Looks like this event shouldn't be triggered if we hit the uniqueness error.

File: controller/schema.go (hunk 22 above)
Comment (jvatic) [inline controller/schema.go]: `event_id` is an int, the append converts it to a string which is necessary. This should probably use a more obvious method of conversion though.

File: controller/resource.go [hunk 24]
```diff
@@ -58,7 +58,16 @@ func (rr *ResourceRepo) Add(r *ct.Resource) error {
 		}
 		r.Apps[i] = postgres.CleanUUID(r.Apps[i])
 	}
-	r.ID = postgres.CleanUUID(r.ID)
+	for _, appID := range r.Apps {
+		if err := createEvent(tx.Exec, &ct.Event{
+			AppID:      appID,
+			ObjectID:   r.ProviderID + ":" + r.ID,
```
Comment (titanous) [inline controller/resource.go]: The resource ID should be unique.

File: controller/schema.go (hunk 22 above)
Comment (titanous) [inline controller/schema.go]: Yeah, use a type coercion thing.

**Comment by jvatic** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
The commits need some body to describe a bit about what's going on, but other than that this LGTM. :shipit:

**Pull request merged**